src/protocol/payload_factory.rs

```diff
@@ -1,29 +1,80 @@
-use crate::protocol::codecs::payload::Payload;
-
-/// A factory for creating payloads.
-#[derive(Clone)]
-pub struct PayloadFactory {
-    payload: Payload,
-}
-
-impl PayloadFactory {
-    /// Create a new payload factory using specified payload as a template.
-    pub fn new(payload: Payload) -> Self {
-        Self { payload }
-    }
-
-    // TODO[asmie]: Add generate() method to start generating payloads (first message).
-    // TODO[asmie]: Re-think API of the PayloadFactory to handle all generic cases (like creating custom payloads with custom fields).
-
-    /// Create a new payload with the same type as the template. If there is a need to
-    /// change any payload fields it's done here.
-    pub fn generate_next(&mut self) -> Payload {
-        // For now, we're incrementing nonce in UniEnsBlockReq and not change other
-        // type of messages.
-        if let Payload::UniEnsBlockReq(message) = &mut self.payload {
-            message.nonce += 1;
-        }
-
-        self.payload.clone()
-    }
-}
+use crate::protocol::codecs::payload::Payload;
+
+/// A factory for creating payloads.
+#[derive(Clone)]
+pub struct PayloadFactory {
+    payload: Payload,
+    customize_payload: fn(&mut Payload) -> (),
+    pre_generated_cache: Vec<Payload>,
+}
+
+impl PayloadFactory {
+    /// Create a new payload factory using specified payload as a template and a function to customize
+    /// the payload. If no customization is needed, just pass None and use default customizer.
+    pub fn new(payload: Payload, customize_payload: Option<fn(&mut Payload) -> ()>) -> Self {
+        let default_customize_payload = |msg: &mut Payload| match msg {
+            Payload::UniEnsBlockReq(message) => {
+                message.nonce += 1;
+            }
+            Payload::UniCatchupReq(message) => {
+                message.nonce += 1;
+            }
+            _ => {}
+        };
+
+        Self {
+            payload,
+            customize_payload: customize_payload.unwrap_or(default_customize_payload),
+            pre_generated_cache: Vec::new(),
+        }
+    }
+
+    /// Create a new payload with the same type as the template. If there is a need to
+    /// change any payload fields customizer is run.
+    pub fn generate_next(&mut self) -> Payload {
+        (self.customize_payload)(&mut self.payload);
+        self.payload.clone()
+    }
+
+    /// Create a new payload with the same type as the template and store it in the cache.
+    pub fn pre_generate_payloads_cache(&mut self, count: usize) {
+        self.pre_generated_cache = self.generate_payloads(count);
+    }
+
+    /// Get vector of payloads reference from the cache.
+    pub fn get_pre_generated_payload_cache(&self) -> &Vec<Payload> {
+        &self.pre_generated_cache
+    }
+
+    /// Generate vector of payloads and return it immediately.
+    pub fn generate_payloads(&mut self, count: usize) -> Vec<Payload> {
+        (0..count).map(|_| self.generate_next()).collect()
+    }
```

Comment:
This could also remain `private` if we consider that the main point of PayloadFactory is to generate payloads as easily/quickly as possible (for the user), where the PayloadFactory takes care of internally buffering the `cached` payloads - all the user has to do is call `generate_next` to fetch a new payload.

Exposing this functionality to the user makes it possible for the user to use two different solutions to the same problem.

approach one:
```rust
let mut factory = PayloadFactory::new(...) // call once
factory.pre_generate_payloads_cache(big_n) // call once
let precious_payload = factory.generate_next() // call whenever needed
```
approach two (dummy example):
```rust
let factory = PayloadFactory::new(...) // call once
let my_local_payloads = factory.generate_payloads(n);

while let Some(payload) = my_local_payloads {
   // use payload
} else {
  my_local_payloads = factory.generate_payloads(new_n);
  continue;
}
```
This is also fine, but then we have two conflicting use cases for the same module here.
I guess both are fine in principle, but maybe we should have a single (idiomatic) solution

Reply:
Maybe we need both options available (`generate_next` and `generate_payloads`), but in that case, those `cache` functions are not really needed here since the second approach returns at once a generated array of payloads

Reply:
Ok, let's start without the cache API and see whether it will be needed.